### ISSUE TITLE
PHP: support for code snippets without the starting `<?php` token

### ISSUE THREAD

**Issue by fabiospampinato** [microsoft/monaco-editor#1658]
**monaco-editor version:** 0.18.1

**Steps or JS usage snippet reproducing the issue:**

The following snippet doesn't get highlighted properly because the starting `<?php` token is missing, I think it should get highlighted better, like GitHub is doing, even though the current behavior is probably overall more correct.

```php
public function index() {
    $projects = Project::all();

    return view('projects.index', compact('projects'));
}
```

**Comment by alexdima** [comment]
The PHP grammar assumes that everything outside `<?`/`<?php`/`?>` is HTML. If we would assume directly PHP, without entering PHP, it would break the coloring of all existing PHP files... We would almost need a new language "php-direct" or something...

**Comment by fabiospampinato** [comment]
> We would almost need a new language "php-direct" or something...

Yeah maybe just a version of the php grammar where the starting `<?`/`<?php` token is optional. I'd call it `php-snippet` 🤔.

It looks like you guys have actually solved this already in vscode, the first editor below contains Markdown and the other two contain PHP, it seems like the starting token is already optional there when embedding PHP inside Markdown:

<img width="864" alt="Screen Shot 2019-12-11 at 15 01 30" src="https://user-images.githubusercontent.com/1812093/70632841-7c201100-1c27-11ea-9611-7424ef5e2b29.png">

**Comment by alexdima** [comment]
TM grammars allow to include named rules from a grammar, so in this case I believe the markdown TM grammar includes directly the inner PHP rules.

For the Monarch PHP parser, the inner PHP is here -- https://github.com/microsoft/monaco-languages/blob/0ed9a6c3e90a24375fab54f7205fb76ce992f117/src/php/php.ts#L204 . AFAIK currently Monarch cannot embed a specific state from within a grammar, so I think we would need a separate php language like you suggest, "php-snippet".

**Comment by billyct** [comment]
I separate the code from [php.ts](https://github.com/microsoft/monaco-languages/blob/main/src/php/php.ts) and register a custom language as `php-snippet`, then it works for me.  

checkout the source code in my app [registerPHPSnippetLanguage.js](https://github.com/tinkerun/tinkerun/blob/master/packages/tinkerun-renderer/src/utils/registerPHPSnippetLanguage.js)

hopes it can help you
